rust/theoros/src/handlers/websocket/subscribe_to_calldata.rs

```diff
@@ -0,0 +1,354 @@
+
+    let id = ws_state.subscriber_counter.fetch_add(1, Ordering::SeqCst);
+
+    let mut subscriber = Subscriber::new(id, Arc::new(state.clone()), feeds_receiver, receiver, sender);
+
+    subscriber.run().await;
+}
+
+pub type SubscriberId = usize;
+
+pub struct Subscriber {
+    id: SubscriberId,
+    closed: bool,
+    state: Arc<AppState>,
+    feeds_receiver: Receiver<CheckpointMatchEvent>,
+    receiver: SplitStream<WebSocket>,
+    sender: SplitSink<WebSocket, Message>,
+    data_feeds_with_config: HashMap<String, DataFeedClientConfig>,
+    active_chain: EvmChainName,
+    ping_interval: tokio::time::Interval,
+    exit: watch::Receiver<bool>,
+    responded_to_ping: bool,
+}
+
+impl Subscriber {
+    pub fn new(
+        id: SubscriberId,
+        state: Arc<AppState>,
+        feeds_receiver: Receiver<CheckpointMatchEvent>,
+        receiver: SplitStream<WebSocket>,
+        sender: SplitSink<WebSocket, Message>,
+    ) -> Self {
+        Self {
+            id,
+            closed: false,
+            state,
+            feeds_receiver,
+            receiver,
+            sender,
+            data_feeds_with_config: HashMap::new(),
+            active_chain: DEFAULT_ACTIVE_CHAIN,
+            ping_interval: tokio::time::interval(PING_INTERVAL_DURATION),
+            exit: crate::EXIT.subscribe(),
+            responded_to_ping: true,
+        }
+    }
+
+    #[tracing::instrument(skip(self))]
+    pub async fn run(&mut self) {
+        while !self.closed {
+            if let Err(e) = self.handle_next().await {
+                tracing::debug!(subscriber = self.id, error = ?e, "Error Handling Subscriber Message.");
+                break;
+            }
+        }
+    }
+
+    async fn handle_next(&mut self) -> Result<()> {
+        tokio::select! {
+            maybe_update_feeds_event = self.feeds_receiver.recv() => {
+                match maybe_update_feeds_event {
+                    Ok(event) => self.handle_data_feeds_update(event).await,
+                    Err(e) => Err(anyhow!("Failed to receive update from store: {:?}", e)),
+                }
+            },
+            maybe_message_or_err = self.receiver.next() => {
+                self.handle_client_message(
+                    maybe_message_or_err.ok_or(anyhow!("Client channel is closed"))??
+                ).await
+            },
+            _  = self.ping_interval.tick() => {
+                if !self.responded_to_ping {
+                    // self.metrics
+                    //     .interactions
+                    //     .get_or_create(&Labels {
+                    //         interaction: Interaction::ClientHeartbeat,
+                    //         status: Status::Error,
+                    //     })
+                    //     .inc();
+
+                    return Err(anyhow!("Subscriber did not respond to ping. Closing connection."));
+                }
+                self.responded_to_ping = false;
```

Comment:
no here we just reset the ping response to false at every interval to make sure that the client has to ping regularly to signal he's still alive otherwise the connection is closed as per the code above